tpe/plugin/src/ShapeFeatures.cc

```diff
@@ -0,0 +1,309 @@
+}
+
+/////////////////////////////////////////////////
+Identity ShapeFeatures::CastToMeshShape(
+  const Identity &_shapeID) const
+{
+  auto it = this->collisions.find(_shapeID);
+  if (it != this->collisions.end() && it->second != nullptr)
+  {
+    auto *shape = it->second->collision->GetShape();
+    if (shape != nullptr && dynamic_cast<tpelib::MeshShape*>(shape))
+      return this->GenerateIdentity(_shapeID, it->second);
+  }
+  return this->GenerateInvalidId();
+}
+
+/////////////////////////////////////////////////
+LinearVector3d ShapeFeatures::GetMeshShapeSize(
+  const Identity &_meshID) const
+{
+  auto it = this->collisions.find(_meshID);
+  if (it != this->collisions.end() && it->second != nullptr)
+  {
+    auto *shape = it->second->collision->GetShape();
+    if (shape != nullptr)
+    {
+      auto *mesh = static_cast<tpelib::MeshShape*>(shape);
+      return math::eigen3::convert(mesh->GetBoundingBox().Size());
+    }
+  }
+  // return invalid size if collision not found
+  return math::eigen3::convert(math::Vector3d(-1.0, -1.0, -1.0));
+}
+
+/////////////////////////////////////////////////
+LinearVector3d ShapeFeatures::GetMeshShapeScale(
+  const Identity &_meshID) const
+{
+  auto it = this->collisions.find(_meshID);
+  if (it != this->collisions.end() && it->second != nullptr)
+  {
+    auto *shape = it->second->collision->GetShape();
+    if (shape != nullptr)
+    {
+      auto *mesh = static_cast<tpelib::MeshShape*>(shape);
+      return math::eigen3::convert(mesh->GetScale());
+    }
+  }
+  // return invalid scale if collision not found
+  return math::eigen3::convert(math::Vector3d(-1.0, -1.0, -1.0));
+}
+
+/////////////////////////////////////////////////
+Identity ShapeFeatures::AttachMeshShape(
+  const Identity &_linkID,
+  const std::string &_name,
+  const ignition::common::Mesh &_mesh,
+  const Pose3d &_pose,
+  const LinearVector3d &_scale)
+{
+  auto it = this->links.find(_linkID);
+  if (it != this->links.end() && it->second != nullptr)
+  {
+    auto &collision = static_cast<tpelib::Collision&>(
+      it->second->link->AddCollision());
+    collision.SetName(_name);
+    collision.SetPose(math::eigen3::convert(_pose));
+
+    tpelib::MeshShape mesh;
+    mesh.SetMesh(_mesh);
+    mesh.SetScale(math::eigen3::convert(_scale));
+    collision.SetShape(mesh);
+
+    return this->AddCollision(_linkID, collision);
+  }
+  return this->GenerateInvalidId();
+}
+
+///////////////////////////////////////////////
+AlignedBox3d ShapeFeatures::GetShapeAxisAlignedBoundingBox(
+  const Identity &_shapeID) const
+{
+  auto it = this->collisions.find(_shapeID);
+  if (it != this->collisions.end() && it->second != nullptr)
+  {
+    auto *shape = it->second->collision->GetShape();
+    if (shape != nullptr)
+      return math::eigen3::convert(shape->GetBoundingBox());
+  }
+  // return invalid bounding box if collision not found
+  return math::eigen3::convert(
+    math::AxisAlignedBox(
```

Comment:
how about just `math::AxisAlignedBox()` with its default [min / max](https://github.com/ignitionrobotics/ign-math/blob/master/src/AxisAlignedBox.cc#L27) values. We have been using that to represent an invalid bounding box